File: libs/fs/CoreCrypto/CoreHash.fs
(* Copyright (C) 2012--2015 Microsoft Research and INRIA *)

module CoreHash
open Bytes

open CryptoProvider

(* ---------------------------------------------------------------------- *)
type engine = HashEngine of MessageDigest

let name (HashEngine engine) =
    engine.Name

let digest (HashEngine engine) (b : bytes) =
    abytes (engine.Digest (cbytes b))

(* ---------------------------------------------------------------------- *)
let md5engine    () = HashEngine (CoreCrypto.Digest "MD5"   )
let sha1engine   () = HashEngine (CoreCrypto.Digest "SHA1"  )
let sha256engine () = HashEngine (CoreCrypto.Digest "SHA256")
let sha384engine () = HashEngine (CoreCrypto.Digest "SHA384")
let sha512engine () = HashEngine (CoreCrypto.Digest "SHA512")

(* ---------------------------------------------------------------------- *)
let dohash (factory : unit -> engine) (x : bytes) =
    let engine = factory () in
        (digest engine x)

let md5    x = dohash md5engine    x
let sha1   x = dohash sha1engine   x
let sha256 x = dohash sha256engine x
let sha384 x = dohash sha384engine x
let sha512 x = dohash sha512engine x

File: libs/ml/DEPRECATED_CoreCrypto/CoreHash.ml
(* -------------------------------------------------------------------- *)
open Bytes
 
type engine = Engine of string * Evp.MD.md

(* -------------------------------------------------------------------- *)
let name (Engine (name, m)) =
  name

let digest (Engine (name, m)) b = 
  let e = Evp.MD.create m in
  Evp.MD.update e (cbytes b);
  let h = Evp.MD.final e in
  Evp.MD.fini e; abytes h

(* -------------------------------------------------------------------- *)
let md5engine    () = Engine ("MD5"   , Evp.MD.md5    ())
let sha1engine   () = Engine ("SHA1"  , Evp.MD.sha1   ())
let sha256engine () = Engine ("SHA256", Evp.MD.sha256 ())
let sha384engine () = Engine ("SHA384", Evp.MD.sha384 ())
let sha512engine () = Engine ("SHA512", Evp.MD.sha512 ())

(* -------------------------------------------------------------------- *)
let md5    b = digest (md5engine    ()) b
let sha1   b = digest (sha1engine   ()) b
let sha256 b = digest (sha256engine ()) b
let sha384 b = digest (sha384engine ()) b
let sha512 b = digest (sha512engine ()) b
